subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyCollector.java

```diff
@@ -166,6 +178,59 @@ public <D extends Dependency> void add(Provider<? extends D> dependency, Action<
         doAddLazy(dependency, configuration);
     }
 
+
+    private static DependencyConstraint applyConstraintConfiguration(DependencyConstraint dependencyConstraint, @Nullable Action<? super DependencyConstraint> config) {
+        if (config != null) {
+            config.execute(dependencyConstraint);
+        }
+
+        // TODO DependencyConstraint mutation validator?
+
+        return dependencyConstraint;
+    }
+
+    private void doAddConstraintEager(DependencyConstraint dependencyConstraint, @Nullable Action<? super DependencyConstraint> config) {
+        getDependencyConstraints().add(applyConstraintConfiguration(dependencyConstraint, config));
+    }
+
+    private void doAddConstraintLazy(Provider<? extends DependencyConstraint> dependencyConstraint, @Nullable Action<? super DependencyConstraint> config) {
+        Provider<DependencyConstraint> provider = dependencyConstraint.map((Object dep) -> {
+            // Generic failure check (for Groovy which ignores this when dynamic)
```

Comment:
❓ How do we get here?  Someone calls `addConstraint(Provider)` from Groovy with the wrong type in the Provider?

Reply:
This seems weird to me as well.  Why are we making a special check here to throw the custom exception?  The default `ClassCastException` will contain the name of the type they are trying to convert, and they should be able to figure out the arguments they are trying to supply are for an overload that doesn't exist.